Improve performance of AggregatingMergeTree w/ SimpleAggregateFunction(String) in PK

azat opened the pull request:
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve performance of AggregatingMergeTree w/ SimpleAggregateFunction(String) in PK

Detailed description / Documentation draft:
While reading from AggregatingMergeTree with
SimpleAggregateFunction(String) in primary key and
optimize_aggregation_in_order perf top shows:

    Samples: 1M of event 'cycles', 4000 Hz, Event count (approx.): 287759760270 lost: 0/0 drop: 0/0
      Children      Self  Shared Object         Symbol
    +   12.64%    11.39%  clickhouse            [.] memcpy
    +    9.08%     0.23%  [unknown]             [.] 0000000000000000
    +    8.45%     8.40%  clickhouse            [.] ProfileEvents::increment   # <-- this, and in debug it has not 0.08x overhead, but 5.8x overhead
    +    7.68%     7.67%  clickhouse            [.] LZ4_compress_fast_extState
    +    5.29%     5.22%  clickhouse            [.] DB::IAggregateFunctionHelper<DB::AggregateFunctionNullUnary<true, true> >::addFree

The reason is obvious, ProfileEvents is atomic counters (and also they
are nested):

<details>

```
    Samples: 7M of event 'cycles', 4000 Hz, Event count (approx.): 450726149337
    ProfileEvents::increment  /usr/bin/clickhouse [Percent: local period]
    Percent│
           │
           │
           │    Disassembly of section .text:
           │
           │    00000000078d8900 <ProfileEvents::increment(unsigned long, unsigned long)@@Base>:
           │    ProfileEvents::increment(unsigned long, unsigned long):
      0.17 │      push  %rbp
      0.00 │      mov   %rsi,%rbp
      0.04 │      push  %rbx
      0.20 │      mov   %rdi,%rbx
      0.17 │      sub   $0x8,%rsp
      0.26 │    → callq DB::CurrentThread::getProfileEvents
           │    ProfileEvents::Counters::increment(unsigned long, unsigned long):
      0.00 │      lea   0x0(,%rbx,8),%rdi
      0.05 │      nop
           │    unsigned long std::__1::__cxx_atomic_fetch_add<unsigned long, unsigned long>(std::__1::__cxx_atomic_base_impl<unsigned long>*, unsigned long, std::__1::memory_order):
      1.02 │      mov   (%rax),%rdx
     97.04 │      lock  add   %rbp,(%rdx,%rdi,1)
           │    ProfileEvents::Counters::increment(unsigned long, unsigned long):
      0.21 │      mov   0x10(%rax),%rax
      0.04 │      test  %rax,%rax
      0.00 │    → jne   78d8920 <ProfileEvents::increment(unsigned long, unsigned long)@@Base+0x20>
           │    ProfileEvents::increment(unsigned long, unsigned long):
      0.38 │      add   $0x8,%rsp
      0.00 │      pop   %rbx
      0.04 │      pop   %rbp
      0.38 │    ← retq
```

</details>

These ProfileEvents was ArenaAllocChunks (it shows ~1.5M events per
second), and the reason is that the table has
SimpleAggregateFunction(String) in PK, which requires Arena.
But most of the time there Arena wasn't even used, so avoid this cost by
re-creating Arena only if it was "used" (i.e. has new chunks).

Another possibility is to avoid populating Arena::head in ctor, but this
will make the Arena code more complex, so for now this was preferred.

~*Also as a long-term solution it worth looking at implementing them via
RCU (to move the extra overhead out from the write code path into read
side).*~

azat committed:
Improve performance of AggregatingMergeTree w/ SimpleAggregateFunction(String)

While reading from AggregatingMergeTree with
SimpleAggregateFunction(String) in primary key and
optimize_aggregation_in_order perf top shows:

    Samples: 1M of event 'cycles', 4000 Hz, Event count (approx.): 287759760270 lost: 0/0 drop: 0/0
      Children      Self  Shared Object         Symbol
    +   12.64%    11.39%  clickhouse            [.] memcpy
    +    9.08%     0.23%  [unknown]             [.] 0000000000000000
    +    8.45%     8.40%  clickhouse            [.] ProfileEvents::increment    # <-- this, and in debug it has not 0.08x overhead, but 5.8x overhead
    +    7.68%     7.67%  clickhouse            [.] LZ4_compress_fast_extState
    +    5.29%     5.22%  clickhouse            [.] DB::IAggregateFunctionHelper<DB::AggregateFunctionNullUnary<true, true> >::addFree

The reason is obvious, ProfileEvents is atomic counters (and also they
are nested):

<details>

```
    Samples: 7M of event 'cycles', 4000 Hz, Event count (approx.): 450726149337
    ProfileEvents::increment  /usr/bin/clickhouse [Percent: local period]
    Percent│
           │
           │
           │    Disassembly of section .text:
           │
           │    00000000078d8900 <ProfileEvents::increment(unsigned long, unsigned long)@@Base>:
           │    ProfileEvents::increment(unsigned long, unsigned long):
      0.17 │      push  %rbp
      0.00 │      mov   %rsi,%rbp
      0.04 │      push  %rbx
      0.20 │      mov   %rdi,%rbx
      0.17 │      sub   $0x8,%rsp
      0.26 │    → callq DB::CurrentThread::getProfileEvents
           │    ProfileEvents::Counters::increment(unsigned long, unsigned long):
      0.00 │      lea   0x0(,%rbx,8),%rdi
      0.05 │      nop
           │    unsigned long std::__1::__cxx_atomic_fetch_add<unsigned long, unsigned long>(std::__1::__cxx_atomic_base_impl<unsigned long>*, unsigned long, std::__1::memory_order):
      1.02 │      mov   (%rax),%rdx
     97.04 │      lock  add   %rbp,(%rdx,%rdi,1)
           │    ProfileEvents::Counters::increment(unsigned long, unsigned long):
      0.21 │      mov   0x10(%rax),%rax
      0.04 │      test  %rax,%rax
      0.00 │    → jne   78d8920 <ProfileEvents::increment(unsigned long, unsigned long)@@Base+0x20>
           │    ProfileEvents::increment(unsigned long, unsigned long):
      0.38 │      add   $0x8,%rsp
      0.00 │      pop   %rbx
      0.04 │      pop   %rbp
      0.38 │    ← retq
```

</details>

These ProfileEvents was ArenaAllocChunks (it shows ~1.5M events per
second), and the reason is that the table has
SimpleAggregateFunction(String) in PK, which requires Arena.
But most of the time there Arena wasn't even used, so avoid this cost by
re-creating Arena only if it was "used" (i.e. has new chunks).

Another possibility is to avoid populating Arena::head in ctor, but this
will make the Arena code more complex, so for now this was preferred.

Also as a long-term solution it worth looking at implementing them via
RCU (to move the extra overhead out from the write code path into read
side).

azat commented:
1.811x faster now

azat commented:
This is very specific use case, but still I will mark it as non draft

azat commented:
>The test 'aggregating_merge_tree_simple_aggregate_function_string' is too slow to run as a whole. Investigate whether the create and fill queries can be sped up

@akuzm how much it is OK for test to run? Not more then 40 seconds?

akuzm commented:
> > The test 'aggregating_merge_tree_simple_aggregate_function_string' is too slow to run as a whole. Investigate whether the create and fill queries can be sped up
> 
> @akuzm how much it is OK for test to run? Not more then 40 seconds?

Something compilcated like "no more than 60 seconds of wall clock time per 13 runs of query (like in master), including all the setup". Normally when "wall clock time" is red and "longest query" is not, this means that the setup phase is too long. But decreasing the query run time will also help (e.g. make it 0.5 second).

akuzm commented:
Btw `system stop merges` or `optimize table` help to get more stable results when you're using a newly created MergeTree table (no background merges when the measured query runs start).

azat committed:
Tune aggregating_merge_tree_simple_aggregate_function_string to make it faster

azat commented:
>Also as a long-term solution it worth looking at implementing them via
>RCU (to move the extra overhead out from the write code path into read
>side).

Does not worth it (and why would, since the write-path is hot, not the read):

```
------------------------------------------------------------------------------
Benchmark                                    Time             CPU   Iterations
------------------------------------------------------------------------------
BM_userspace_rcu                          3.55 ns         3.55 ns    180008939 # bp version
BM_userspace_rcu_uatomic_add              6.94 ns         6.94 ns    100860254 # bp version
BM_userspace_rcu_uatomic_add_return       9.64 ns         9.64 ns     72620501 # bp version
BM_atomic_empty                           2.05 ns         2.05 ns    293076586
BM_atomic_load_relaxed                    2.07 ns         2.07 ns    338530084
BM_atomic_load_acquire                    2.10 ns         2.10 ns    333778097
BM_atomic_store_release                   2.02 ns         2.02 ns    347426790
BM_atomic_store_seq_cst                   9.64 ns         9.64 ns     72618247
BM_atomic_fetch_add_relaxed               9.64 ns         9.64 ns     72614481
BM_atomic_fetch_add_seq_cst               9.71 ns         9.70 ns     72455484
BM_atomic_acquire_fence                   2.10 ns         2.10 ns    333774984
BM_atomic_seq_cst_fence                   27.0 ns         27.0 ns     25935192
BM_atomic_fetch_add_cs                    39.1 ns         39.0 ns     17922906
```

https://gist.github.com/azat/066d165154fa1efe6000fac59062cc25#file-summary-txt

akuzm committed:
Update aggregating_merge_tree_simple_aggregate_function_string.xml

`system stop merges` w/o table name has global effect, so the rest of the tests is affected. Also `optimize` is more suitable here so that the end result is the same every time.

akuzm commented on src/Processors/Merges/Algorithms/AggregatingSortedAlgorithm.cpp:
```diff
@@ -212,8 +219,17 @@ void AggregatingSortedAlgorithm::AggregatingMergedData::startGroup(const ColumnR
         desc.createState();
 
-    if (def.allocates_memory_in_arena)
+    /// If and only if:
+    /// - arena is required (i.e. SimpleAggregateFunction(any, String) in PK)
+    /// - arena was used since otherwise it may be too costly to increment atomic counters inside Arena.
+    ///   i.e. SELECT with SimpleAggregateFunction(String) in PK and lots of groups
+    ///   may produce ~1.5M of ArenaAllocChunks atomic increments,
+    ///   while LOCK is too costly for CPU (~10% overhead here).
```
```suggestion
    /// Frequent Arena creation may be too costly, because we have to increment the atomic
    /// ProfileEvents counters when creating the first Chunk -- e.g. SELECT with
    /// SimpleAggregateFunction(String) in PK and lots of groups may produce ~1.5M of
    /// ArenaAllocChunks atomic increments, while LOCK is too costly for CPU
    /// (~10% overhead here).
    /// To avoid this, reset arena if and only if:
    /// - arena is required (i.e. SimpleAggregateFunction(any, String) in PK),
    /// - arena was used in the previous groups.
```

akuzm reviewed (approved):
LGTM

nvartolomei commented:
Slightly off-topic: Why would you use a SimpleAggregateFunction as part of the primary key?

azat committed:
Better description of costly ArenaAllocChunks in AggregatingSortedAlgorithm

Co-authored-by: Alexander Kuzmenkov <36882414+akuzm@users.noreply.github.com>

azat commented:
>Slightly off-topic: Why/when would you use a SimpleAggregateFunction as part of the primary key?

Good question! And the answer is pretty simple - old poor design
Actually it worth forbid such use cases (as an addition for #16601)

akuzm commented:
Test failures don't seem to be relevant.
